Implementation/Custom-Routing/signalR.md

```diff
@@ -0,0 +1,138 @@
+---
+versionFrom: 9.0.0
+product: "CMS"
+meta.Title: "Adding a signlR hub"
+meta.Description: "Umbraco ships with signalR installed, find out how to add your own hub(s) to the existing setup"
+
+---
+
+# Todo 
+- Create a hub
+- Define a custom route for it
+- Add the routing to the umbraco composer
+- Test the setup with some javascript
+
+## Create a hub
+We are going to go for the most simple implementation possible, a simple status ping. So create a new class with the following code
+```csharp
+using Microsoft.AspNetCore.SignalR;
+
+public class TestHub : Hub
+{
+    public async Task Ping()
+    {
+        await Clients.All.SendAsync("Pong");
+    }
+}
+```
+
+### Define a custom route
+Next up is defining a custom route, for this we are going to use a `IAreaRoutes` and the base umbrace backend path so we dont have to reserve another path in the settings
+```csharp
+using Microsoft.AspNetCore.Builder;
+using Microsoft.Extensions.DependencyInjection;
+using Umbraco.Cms.Core.Composing;
+using Umbraco.Cms.Core.DependencyInjection;
+using Umbraco.Cms.Web.Common.ApplicationBuilder;
+using Umbraco.Extensions;
+
+public class TestHubRoutes : IAreaRoutes
+{
+    private readonly IRuntimeState _runtimeState;
+    private readonly string _umbracoPathSegment;
+
+    public TestHubRoutes(
+        IOptions<GlobalSettings> globalSettings,
+        IHostingEnvironment hostingEnvironment,
+        IRuntimeState runtimeState)
+    {
+        _runtimeState = runtimeState;
+        _umbracoPathSegment = globalSettings.Value.GetUmbracoMvcArea(hostingEnvironment);
+    }
+
+    public void CreateRoutes(IEndpointRouteBuilder endpoints)
+    {
+        switch (_runtimeState.Level)
+        {
+            case Umbraco.Cms.Core.RuntimeLevel.Run:
+                endpoints.MapHub<TestHub>(GetTestHubRoute());
+                break;
+        }
+
+    }
+
+    public string GetTestHubRoute()
+    {
+        return $"/{_umbracoPathSegment}/{nameof(TestHub)}";
+    }
+}
+```
+
+### Add the routing to the umbraco composer
+Last step in the setup is registering our custom route
+
+```csharp
+using Microsoft.AspNetCore.Builder;
+using Microsoft.Extensions.DependencyInjection;
+using Umbraco.Cms.Core.Composing;
+using Umbraco.Cms.Core.DependencyInjection;
+using Umbraco.Cms.Web.Common.ApplicationBuilder;
+using Umbraco.Extensions;
+
+public class TestHubComposer : IComposer
+{
+    public void Compose(IUmbracoBuilder builder)
+    {
+        builder.Services.AddUnique<TestHubRoutes>();
+        builder.Services.AddSignalR();
```

Comment:
I'm in doubt if this should be called.. It is already called by BackOffice (we use Signal-R for the previews). Calling this twice could potentially be a bad thing.

Alternatively, should you check if it is already added. We do something similar in the CMS a couple of places..

Reply:
Hey Bjarke, I have looked around, but I can't find how to check if something has been added to the services yet. Could you help me out?

Reply:
I agree it is never easy with build in stuff, but AddSignalR `seems to be safe to call multiple times, as it uses statements like:
```
services.TryAddSingleton(typeof(IHubContext<>), typeof(HubContext<>));
```

Under the hood, it does something similar to this, but for each of the types, it adds to the container, so it is not super-efficient.

Maybe doing something like this will be the best compromise, as we only iterate the collection once:
 ```
 if (!builder.Services.Any(x => x.ServiceType == typeof(IHubContext<>)))
{
    builder.Services.AddSignalR();
}
```